Added basic scripts to run pyscf.

babbush opened the pull request:
I am adding a basic script to run pyscf and also some example files. A team effort is needed to complete this repo. @damiansteiger and @thomashaener, I will need help from you two in writing the setup scripts. I have run the examples here by manually modifying my pypath. @damiansteiger, I will also need you to add the appropriate license headers to these files. I would have done it myself but since this is a BSD repo, I did not have any examples from you of the correct headers. @jarrodmcc, if you run the example files you will see that pyscf fails to converge for even simple examples, especially for the periodic table. It also fails (embarrassingly) for a couple points on the hydrogen molecule curve. I think we need to change the scf options to something different so that the SCF converges. Or, alternatively, we need a better way to handle errors that come from pyscf. Can you help with this?

Note that some of the example code here is almost identical to example code in FermiLib-Plugins-Psi4. This is intentional because some people will install just one of the plugins and we want them to have these examples. I assume this isn't going to be a problem.

babbush committed:
Added basic scripts to run pyscf. Much is needed before this repo can be released, including proper license headers.

jarrodmcc committed:
Fixed incorrect orbital number reference and disabled CISD and CCSD for non-singlet references as it does not yet appear to be implemented in PySCF

jarrodmcc committed:
Minor pep8 fixes

jarrodmcc reviewed (dismissed):
Looks good as a base commit.  An issue that cannot be resolved by us is that PySCF is currently unable to do CISD and CCSD for open shell ROHF references, meaning no odd numbered electrons.  I added a not implemented warning for this situation, and opened an issue on the PySCF repo in regards to this.

damiansteiger reviewed (changes requested):
I will look into the license.

babbush commented:
Great, thanks Jarrod! Damian, I think all you need to do is put the license header in the folder. Once that is done, let's accept the PR. The next item would be to have you guys help with the install scripts.

babbush commented:
@damiansteiger Damian can you please add the license header and approve this?

damiansteiger commented on fermilibpluginpyscf/_run_pyscf.py:
```diff
@@ -0,0 +1,255 @@
+    return molecule
+
+
+# Test.
```
Tests belong into a separate file

babbush replied:
We cannot really test pyscf without setting it up with coveralls and travis. So I just removed this code.

damiansteiger commented on fermilibpluginpyscf/_run_pyscf.py:
```diff
@@ -0,0 +1,255 @@
+    if 1:
+        molecule = run_pyscf(
+            molecule, run_scf, run_mp2, run_cisd, run_ccsd, run_fci, verbose)
+    else:
```
Dead code

babbush replied:
Removed it.

damiansteiger commented on fermilibpluginpyscf/_run_pyscf.py:
```diff
@@ -0,0 +1,255 @@
+              run_ccsd=False,
+              run_fci=False,
+              verbose=False):
+    """This function runs a Psi4 calculation.
```
Fix docstring

babbush replied:
I fixed the missing run_mp2 option. But I did not add the space if that is what you are asking for. I want all files in FermiLib to be consistent. If you insist on this space then please update all others files. We can do that before moving out of alpha but I don't want to be half in one convention and half in another.

damiansteiger replied:
*) This function runs a **Psi4** calculation...
*) Yes, there needs to be a newline for each function docstring as this is our convention.

babbush replied:
Fixed the psi4 thing. Again, the rest of FermiLib does not have that doc string convention. So apparently, it is not the convention.

damiansteiger commented on examples/plotter.py:
```diff
@@ -0,0 +1,97 @@
+
+with warnings.catch_warnings():
+    warnings.simplefilter('ignore')
+    import pylab
```
Consider importing matplotlib instead of pylab which pollutes the namespace

damiansteiger replied:
see https://stackoverflow.com/questions/16849483/which-is-the-recommended-way-to-plot-matplotlib-or-pylab

babbush replied:
This is also used in the psi4 plugin. Let's change from pylab to matplotlib later, since other parts of FermiLib currently use pylab.

damiansteiger replied:
What warning do you suppress?

babbush replied:
I don't know actually, that is unnecessary code, good catch. I removed it.

damiansteiger reviewed (changes requested):
Thanks for creating these scripts :1st_place_medal: A few things need to change, see comments and:

* Don't commit `examples/.ipynb_checkpoints/pyscf_demo-checkpoint.ipynb`
* Please adapt pyscf_demo.ipynb, I can still find at various places "psi4" instead of "pyscf"
* Please include the standard Apache header of the ProjectQ-Framework. While Pyscf is under BSD license, we can release this code under Apache.

babbush committed:
Added license headers, cleaned up docs, removed __main__ from _run_pyscf.py

babbush commented:
I fixed the major issues you brought up, added the license headers and removed mention of psi4 from the demo. Please accept and we can make other changes later.

damiansteiger commented on examples/pyscf_demo.ipynb:
```diff
@@ -0,0 +1,719 @@
+     "output_type": "stream",
+     "text": [
+      "Warn: Ipython shell catchs sys.args\n",
+      "Warn: Ipython shell catchs sys.args\n"
```
What is this about?

babbush replied:
No idea actually. There are a few things here we need to debug still, clearly. Please let's accept this PR and then fix later. There is a ton of stuff in this PR already.

damiansteiger replied:
The ProjectQ-Framework follows a high coding standard with mandatory code reviews. It would defeat the purpose of mandatory code reviews, if we accept pull requests with unknown warnings (which could then lead to unknown consequences).

If debugging this error takes too long, then please remove this part and submit it with a separate pull request once it has been fixed.

babbush replied:
It is coming from pyscf. It is likely a bug on their end. The code where it is raised is here: http://sunqm.net/pyscf/_modules/gto/mole.html

We need the example code here and it works aside from this warning. Maybe you can help me fix this? This isn't even the module code, its the demo code and it seems to be a conflict with ipython. The code runs when called from normal python.

babbush replied:
I think it is unreasonable to reject a pull request because of a warning caused by a collision between ipython and a third party library which does not effect in any way the calculation. But I would also be happy if you could resolve this. Otherwise, I think you should accept this PR and open an issue about the warning.

damiansteiger replied:
Thanks for looking up where this warning came from. There seems not much we can do about it so I am okay with this warning.

I was just wondering what this warning is about especially as the only purpose of this repo is to interface with this one "third party" library. Sorry for doing due diligence and not accepting this blindly.

babbush committed:
Fixed doc string

babbush committed:
Added unnecessary warning catch

damiansteiger commented on examples/plotter.py:
```diff
@@ -0,0 +1,106 @@
+
+"""These functions compare properties of different molecules."""
+import numpy
+import pylab
```
Please change to matplotlib.

>http://matplotlib.org/faq/usage_faq.html#matplotlib-pyplot-and-pylab-how-are-they-related:
[...]Although many examples use pylab, it is no longer recommended.[...]

babbush replied:
Done. Please do the same for us in the other plugin repo since you felt so strongly.

damiansteiger reviewed (changes requested):
* New line in docstrings for functions.
* Debugging or removing code which generated a warning

babbush commented on examples/plotter.py:
```diff
@@ -0,0 +1,106 @@
+
+"""These functions compare properties of different molecules."""
+import numpy
+import pylab
```
There is no code that generates a warning.

If you add a new line for each doc string in the rest of FermiLib (which all follows a different convention) then I will match the "new" convention here. Otherwise, I am going to be consistent with the rest of the library.

babbush committed:
Moved to matplotlib

damiansteiger commented:
Dear Ryan,

This pull request seems almost ready to go 👍 Thanks for putting your time and effort into this interface for PySCF.

The only remaining issue is the missing new line for the docs strings of functions.

In this new repo, we follow the same coding conventions as the ProjectQ-Framework (for a reference please see the main ProjectQ repo) which uses the convention of a new line for function docstrings. As you feel very strongly about not adding a new line, I will discuss it next week with Thomas if we want to make an exception for this repo. We consider changing coding conventions very seriously, and hence please excuse the extra amount of time it will take for this.

Should this PR be time critical, I would suggest to add the new line in the docstring and discuss changing the coding convention at a later point in time.

babbush committed:
added line for docstring

damiansteiger committed:
Update _run_pyscf.py
